PWGUD/Tasks/diffMCDataScanner.cxx

```diff
@@ -0,0 +1,486 @@
+
+using namespace o2;
+using namespace o2::framework;
+
+void customize(std::vector<ConfigParamSpec>& workflowOptions)
+{
+  workflowOptions.push_back(ConfigParamSpec{"runCase", VariantType::Int, 0, {"runCase: 0 - histos,  1 - mcTruth, else - tree"}});
+}
+
+#include "Framework/runDataProcessing.h"
+#include "Framework/AnalysisTask.h"
+#include "CommonConstants/LHCConstants.h"
+#include "Common/CCDB/TriggerAliases.h"
+#include "Common/DataModel/EventSelection.h"
+#include "Common/DataModel/TrackSelectionTables.h"
+#include "Common/Core/PID/PIDResponse.h"
+#include "Common/Core/MC.h"
+
+using namespace o2::framework::expressions;
+
+namespace o2::aod
+{
+namespace datascan
+{
+DECLARE_SOA_COLUMN(PID, pid, int);   //! MCTruth PID
+DECLARE_SOA_COLUMN(Pt, pt, float);   //! MCTruth particle pt
+DECLARE_SOA_COLUMN(sEL, sel, float); //! TPC nSigma el
+DECLARE_SOA_COLUMN(sMU, smu, float); //! TPC nSigma mu
+DECLARE_SOA_COLUMN(sPI, spi, float); //! TPC nSigma pi
+DECLARE_SOA_COLUMN(sKA, ska, float); //! TPC nSigma ka
+DECLARE_SOA_COLUMN(sPR, spr, float); //! TPC nSigma pr
+} // namespace datascan
+DECLARE_SOA_TABLE(nSigmas, "AOD", "NSIGMAS", //! MCTruth of particle PID and pt and TPC nSigma for el, mu, pi, ka, pr
+                  datascan::PID, datascan::Pt,
+                  datascan::sEL, datascan::sMU, datascan::sPI,
+                  datascan::sKA, datascan::sPR);
+} // namespace o2::aod
+
+template <typename T>
+T getCompatibleBCs(soa::Join<aod::Collisions, aod::EvSels, aod::McCollisionLabels>::iterator const& collision, T const& bcs)
+{
+  LOGF(debug, "Collision time / resolution [ns]: %f / %f", collision.collisionTime(), collision.collisionTimeRes());
+
+  auto bcIter = collision.bc_as<T>();
+
+  // due to the filling scheme the most probably BC may not be the one estimated from the collision time
+  uint64_t mostProbableBC = bcIter.globalBC();
+  uint64_t meanBC = mostProbableBC - std::lround(collision.collisionTime() / o2::constants::lhc::LHCBunchSpacingNS);
+  int deltaBC = std::ceil(collision.collisionTimeRes() / o2::constants::lhc::LHCBunchSpacingNS * 4);
+  int64_t minBC = meanBC - deltaBC;
+  int64_t maxBC = meanBC + deltaBC;
+  if (minBC < 0) {
+    minBC = 0;
+  }
+
+  // find slice of BCs table with BC in [minBC, maxBC]
+  int64_t maxBCId = bcIter.globalIndex();
+  int moveCount = 0; // optimize to avoid to re-create the iterator
+  while (bcIter != bcs.end() && bcIter.globalBC() <= maxBC && bcIter.globalBC() >= minBC) {
+    LOGF(debug, "Table id %d BC %llu", bcIter.globalIndex(), bcIter.globalBC());
+    maxBCId = bcIter.globalIndex();
+    ++bcIter;
+    ++moveCount;
+  }
+
+  bcIter.moveByIndex(-moveCount); // Move back to original position
+  int64_t minBCId = collision.bcId();
+  while (bcIter != bcs.begin() && bcIter.globalBC() <= maxBC && bcIter.globalBC() >= minBC) {
+    LOGF(debug, "Table id %d BC %llu", bcIter.globalIndex(), bcIter.globalBC());
+    minBCId = bcIter.globalIndex();
+    --bcIter;
+  }
+
```

Comment:
I think this code is copied from compatibleBCs.cxx right?
Shouldn't we better move it to a header and use the function in both cxx files?

Reply:
yes indeed, this is an adapted version of compatibleBCs.cxx. For the moment I would like to keep it here to be able to make modifications without having to worry about the impact on other tasks. But I think you are right, that this is a function which could be useful for others as well.